### ISSUE TITLE
DispatchIO.read spinning while pipe is open on Windows

### ISSUE THREAD

**Issue by DimaKozhinov** [swiftlang/swift-corelibs-libdispatch#820]
The sourcekit-lsp.exe consumes ~24% of CPU power all the time, even when idle. This bug does not appear under Linux, and does not appear in Swift versions earlier than v5.9. It is still present in v5.10. I tested this under VirtualBox, my virtual machine has 4 virtual processors. The host machine has 13th Gen Intel Core i7 processor, so ~24% load seems like a lot of mysterious calculations.

How to reproduce:

1. Install Swift version 5.9 .. 5.10. Install Microsoft VS Code and Swift extension for it (v1.8.1 from Swift Server Work Group).
3. Create a test Swift package:
    mkdir test
    cd test
    swift package init
4. Open the test folder in VS Code. (File | Open Folder...). At this point CPU usage increases, you don't even need to edit any source file. On my (usually silent) PC fan immediately starts rotating much faster, producing loud noise.
5. Check CPU usage by the sourcekit-lsp.exe: Press Ctrl-Shift-Esc, in the process list find the Visual Studio Code, click ">" next to it to see its sub-processes, and see high CPU usage by sourcekit-lsp .
6. Return to step 1 and this time install Swift version 5.8.1 or earlier. Repeat step 2 (create Swift package) because earlier version of Swift won't open a package created with a later version. Repeat all next steps and see ~0% CPU usage by sourcekit-lsp.exe .

This bug has nothing to do with VS Code. I found this bug when using my own app that runs sourcekit-lsp.exe, and not VS Code.

**Comment by ahoppen** [comment]
@tristanlabelle I have reached that stage. If you could investigate it, I would really appreciate it.

**Comment by tristanlabelle** [comment]
Hi @ahoppen , we'll look into it. Currently our priorities in this area are roughly:
1. SourceKit-LSP file locking issue
2. https://github.com/apple/sourcekit-lsp/issues/1139
3. This

**Comment by ahoppen** [comment]
It looks like we are spending the time in `_dispatch_event_loop_leave_immediate`. @tristanlabelle Have you seen this before in your previous fixes of hangs in libdispatch?

<img width="742" alt="Screenshot 2024-07-30 at 19 21 13" src="https://github.com/user-attachments/assets/6e0b8b52-10be-4a6c-9574-4b0c3dbdec67">

**Comment by tristanlabelle** [comment]
@ahoppen I don't remember seeing this before specifically. The trace will be useful though as this is next in the priority list. We got through the [file locking issue](https://github.com/swiftlang/swift/pull/75406) and I don't think we're hitting the crash anymore (from my list in previous comment). fyi @z2oh

**Comment by ahoppen** [comment]
That would be amazing. I wonder what kind of fix this will be.

**Comment by z2oh** [comment]
I've figured out what's causing this, but I'm not sure of a proper fix yet.

By default, Windows pipes are blocking on both the read and write side (`PIPE_WAIT`).

libdispatch seems to expect non-blocking write semantics on pipes, and so this was effected on Windows in https://github.com/apple/swift-corelibs-libdispatch/pull/781 by setting the `PIPE_NOWAIT` flag here:
https://github.com/apple/swift-corelibs-libdispatch/blob/85369bae730669dcee18b771f27f885250971900/src/io.c#L1440-L1456

Perhaps surprisingly, this also changes the read side of the pipe to do non-blocking reads. The original Windows pipe-event implementation (https://github.com/apple/swift-corelibs-libdispatch/pull/501) took advantage of Windows' blocking-by-default read semantics and spawned a lightweight thread to call `ReadFile` on the pipe in a loop.

I'm not sure when `ReadFile` on a `PIPE_WAIT` pipe would result in `ERROR_NO_DATA` (the [docs](https://learn.microsoft.com/en-us/windows/win32/ipc/named-pipe-type-read-and-wait-modes?redirectedfrom=MSDN#wait-mode) don't describe this case in detail), but the dispatch code handles this error by restarting the loop. Under `PIPE_NOWAIT`, `ERROR_NO_DATA` is returned immediately on a read of an empty pipe, turning this code into a CPU-hogging polling loop.

https://github.com/apple/swift-corelibs-libdispatch/blob/85369bae730669dcee18b771f27f885250971900/src/event/event_windows.c#L299-L303

The comment in the code here seems to indicate that this error implies we have a non-blocking pipe (which is true as we've seen), but I can't intuit why that should cause the loop to restart. Perhaps there was an assumption that the pipe is still initializing and would eventually be set to `PIPE_WAIT`? I'm not sure.

https://github.com/apple/swift-corelibs-libdispatch/pull/781 was cherry-picked into Swift 5.9 (and after), which explains why this issue isn't present in earlier versions.

I've built a local `dispatch.dll` that switches the pipe back to `PIPE_WAIT` and I no longer observe the perpetually spinning thread in sourcekit-lsp or in the test repro provided by @ahoppen. Presumably, this reintroduces the bug(s) that https://github.com/apple/swift-corelibs-libdispatch/pull/781 resolved, so the actual fix will need to be more sophisticated. I'll continue to look into this.

**Comment by tristanlabelle** [comment]
It makes sense to me that a final solution would be with `PIPE_WAIT` pipes, especially since `PIPE_NOWAIT` is deprecated. For context on the original issue: a `PIPE_WAIT` write will block if there's not enough buffer space to complete the write, and we never were able to find a reliable way to know that there will not be enough buffer space before attempting the write. The function that is supposed to report that information spuriously reports a misleading value of zero bytes.
Some possible avenues of investigation:
- Can we get the reader side of the pipe from the writer and query how much data is available to read as a proxy to determine how much of the pipe is empty and available to write?
- Can we do an overlapped write operation and then cancel it if it doesn't complete immediately because the pipe is full?

**Comment by litewrap** [comment]
Just for my information, is there any progress on resolving this very annoying bug on Windows and identified 8 months ago?

**Comment by tristanlabelle** [comment]
@litewrap , I can only speak for the Browser Company and we're not scheduling this work right now. Our investigation to this point showed that it will require a refactoring of libdispatch that stretches our familiarity with the codebase. It doesn't mean we won't do it as I agree it's a bad bug, but we're not currently blocked on it.

**Comment by rauhul** [comment]
FWIW any rewrites of parts of corelibs dispatch should consider using swift, I'm currently battling a dispatch segfault linux.

**Comment by Keithcat1** [comment]
Yeah, just ran into this on Swift 6.0.2, Windows 11 64-bit. Simply leaving VSCode open will make my laptop fan go crazy.

**Comment by z2oh** [comment]
I had some more time to look into this, and I think I have a fix!

To summarize the problem:
* In https://github.com/swiftlang/swift-corelibs-libdispatch/pull/781, the pipe was changed from `PIPE_WAIT` to `PIPE_NOWAIT` to achieve POSIX-like `O_NONBLOCK` semantics, as attempting to write a large buffer would cause the dispatch queue thread to block long enough to hit a time out.
* The Windows pipe synchronization thread, `_dispatch_pipe_monitor_thread`, uses a blocking 0-byte read as a synchronization mechanism to signal to the actual reader threads that there is data waiting in the pipe. Of course, for this to work, the read must actually be blocking. With a `PIPE_NOWAIT` pipe, the Read will not block, causing the thread to spin endlessly and constantly wake the actual reader thread to perform 0 byte reads. This spinning thread is the root source of the problem in this issue.

Switching the pipe back to `PIPE_WAIT` is easy enough, but then the original problem of blocking writes resurfaces. However, after a lot of investigation into Windows pipe semantics, I think https://github.com/swiftlang/swift-corelibs-libdispatch/pull/796 already fixes the blocking problem (this was a follow-up PR to fix a problem with the `PIPE_NOWAIT` implementation), which would mean the simple change back to `PIPE_WAIT` is all that's needed.

Bounding the size of the write to the size of the buffer ensures we don't block for too long on the dispatch queue:
https://github.com/swiftlang/swift-corelibs-libdispatch/blob/0c38954e0033f4ef112fc2b6d8244c798dce30da/src/io.c#L2516-L2528
* We first check the `WriteQuotaAvailable` of the pipe (which is the size of the pipe's buffer minus the total amount of space queued to be read), and use this as an initial upper bound.
* If `WriteQuotaAvailable == 0`, this implies we have a reader that has requested a full buffer's worth of data from the pipe, and so we bound by the size of the pipe's buffer (`OutboundQuota`).
* Either that reader is currently draining the pipe, in which case writing the full size of the buffer will be able to finish quickly after the pipe is drained
* _or_ the pipe is currently empty with a blocked reader waiting for data, in which case the full buffer write will _also_ be able to finish quickly.

With the following simple patch:
```
diff --git a/src/io.c b/src/io.c
index e31e28c..38e46b6 100644
--- a/src/io.c
+++ b/src/io.c
@@ -1441,16 +1441,15 @@ _dispatch_fd_entry_create_with_fd(dispatch_fd_t fd, uintptr_t hash)
                                // from Foundation.Pipe may not have FILE_WRITE_ATTRIBUTES.
                                DWORD dwPipeMode = 0;
                                if (GetNamedPipeHandleState((HANDLE)fd, &dwPipeMode, NULL,
-                                               NULL, NULL, NULL, 0) && !(dwPipeMode & PIPE_NOWAIT)) {
-                                       dwPipeMode |= PIPE_NOWAIT;
+                                               NULL, NULL, NULL, 0) && !(dwPipeMode & PIPE_WAIT)) {
+                                       dwPipeMode |= PIPE_WAIT;
                                        if (!SetNamedPipeHandleState((HANDLE)fd, &dwPipeMode,
                                                        NULL, NULL)) {
```
I'm seeing no spinning from sourcekit-lsp or the test program provided above. Anecdotally, sourcekit-lsp feels much snappier and more responsive with this patch in place (though I haven't measured to be sure). I've convinced myself _in theory_ that the original hang condition that precipitated the switch to `PIPE_NOWAIT` won't occur thanks to the bounding as described above. The only thing left to do is to demonstrate this in practice by replicating the hang with `PIPE_WAIT` without the bounding logic in place, and then verifying that the re-adding this logic avoids the hang.

I hope to have a PR up soon!

**Comment by ahoppen** [comment]
Thank you so much for diving deep and investigating this @z2oh 🤩

**Comment by z2oh** [comment]
PR is up!

https://github.com/swiftlang/swift-corelibs-libdispatch/pull/853 just switches the pipe back to `PIPE_WAIT`, but https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 adds an additional trick to enable `WriteQuotaAvailable == 1` to be used as a sentinel value to distinguish the "output buffer is full" case. By always leaving a free byte of space in the buffer when writing, the only way that `WriteQuotaAvailable == 0` can happen is if a reader has requested a full buffer's worth of data, so we can avoid blocking on the `WriteFile` in all cases. There _may_ be issues with this approach, as it means we can no longer serially write-then-read `bufsize` worth of data, only `<bufsize`.

Both of these PRs fix the spinning thread issue though, and if https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 is not mergeable then I think https://github.com/swiftlang/swift-corelibs-libdispatch/pull/853 will do.

**Comment by ahoppen** [comment]
This is great :z2oh: Thank you some much for your investigation and the fix 🎉 🙌🏽

**Comment by litewrap** [comment]
Is this fixed ?

**Comment by z2oh** [comment]
> Is this fixed ?

Not yet. Either #854 or #853 still need to be reviewed and merged. This is a subtle/tricky change with a lot of background context to ingest, so a thorough review is not easy. I'll try and see if I can resurface this on the relevant folks' radar to hopefully get a review soon.

**Comment by DimaKozhinov** [comment]
It's been a year already... The bug is really painful and damaging for Swift ecosystem on Windows.

**Comment by compnerd** [comment]
@DimaKozhinov you are welcome to take up the efforts to solve the issue. We have tried a few different approaches, but they all find different ways that they are insufficient.

**Comment by JoaoVictorVP** [comment]
Can't we like put a small sleep on the thread while it is spinning as a short-time solution, I think the drawbacks of this are less likely to be worse than consuming constantly 15% of CPU and making the experience of whoever is developing laggier because of it

**Comment by z2oh** [comment]
We had a chat with a Win32 pipes maintainer at Microsoft and it was confirmed that there is no way to reliably inspect the free space of a pipe's outbound buffer (even in a single-threaded context). The guidance offered was to use worker threads on the write side to emulate asynchronous writing. This solution is not intractable, although it introduces complexity and adds additional thread pressure to the system. I'm not sure when I will be able to dedicate time in pursuing this solution (though I may be able to at some point in the future).

In the meantime, I still think etiher https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 or https://github.com/swiftlang/swift-corelibs-libdispatch/pull/853 is worth merging. These solutions are not flawless: they trade off the perpetually spinning thread for a tiny bit of correctness, where under very specific conditions it is possible for a write to never make progress (importantly: without blocking the dispatch queue). But in the weeks that I've been using this patched version of libdispatch, I have not able to produce these conditions in practice, even with synthetic pathological experiments.

**Comment by ahoppen** [comment]
Thank you for continuing to investigate this, @z2oh and checking with the Microsoft engineers 🙏🏽 

What’s your opinion on which is the better solution to fix this problem? https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 or https://github.com/swiftlang/swift-corelibs-libdispatch/pull/853? Also, your comment here sounds like there is a correctness issue with both approaches but the description of https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 mentions that it doesn’t have the correctness issue #853. I just want to make sure that you didn’t make any new discoveries that aren’t reflected in the PR descriptions.

Also, one question about the explanation in #853
> So if nothing is actively reading the data, the queue will be blocked from making progress. However, I have been unable to trigger this case in practice. If there is no reader on the pipe, `WriteFile` will fail quickly. From what I could grok from the libdispatch code, the pipe synchronization mechanism will always have a reader waiting on some I/O Completion Port that will quickly resume when data in the pipe is available to be read.

I might be misunderstanding what you mean by *if nothing is actively reading the data*. Couldn’t the other end of the pipe be a separate process that just never reads data from the pipe. Since that process might be implemented without libdispatch, I’m not sure how assumptions about the libdispatch reader can affect the writer.

**Comment by z2oh** [comment]
There are no new discoveries, but information is a bit scattered between the two PRs and this issue thread. I will attempt to collate and clarify here.

### Summary

There are two potential correctness issue I'm thinking about:
1. The dispatch queue blocks (permanently or long enough to hit timeouts) on the call to `WriteFile`. 
2. Data passed into libdispatch is not written through to the pipe, despite a reader attempting to read data. _N.B._ In the absence of a reader, I would consider it not incorrect to avoid writing data.

The first type of incorrectness was present in libdispatch >=2 years ago. I can't find the issue now, but I recall that this was directly affecting users of sourcekit-lsp with crashes/stalls on large-completion buffers (which would attempt to write the full buffer in a single synchronous call, blocking the queue until all data was read or a timeout was encountered), and was fixed by https://github.com/swiftlang/swift-corelibs-libdispatch/pull/781 which switched the pipe to `PIPE_NOWAIT` to ensure that `WriteFile` no longer blocked. However, this change inadvertently introduced two other bugs:
1. the spinning thread issue due to reads on a `PIPE_NOWAIT` pipe no longer blocking on the monitoring thread
2. the second correctness issue listed above (materialized as https://github.com/swiftlang/sourcekit-lsp/issues/752). A `WriteFile` on a  `PIPE_NOWAIT` pipe [may not write any bytes if the entire write cannot fit into the pipe's buffer](https://learn.microsoft.com/en-us/windows/win32/ipc/named-pipe-type-read-and-wait-modes#wait-mode), meaning we would never make progress writing if the write request exceeded the pipe's outbound buffer size.

The second bug was fixed by a followup PR (https://github.com/swiftlang/swift-corelibs-libdispatch/pull/796) which puts an upper bound on the attempted write. Because the write request is now bounded by the pipe's outbound buffer size, we can be confident that progress will be made as long as there is a reader draining the pipe. So now we are left with the spinning thread issue, and this is the state of things today.

Both #854 and #853 switch the pipe back to `PIPE_WAIT` in order to fix the spinning thread issue. As libdispatch pipes expect `O_NONBLOCK` semantics it was tempting to try and leave `PIPE_NOWAIT` around and instead fix this by changing the pipe monitoring thread mechanism to no longer rely on a blocking read, but `PIPE_NOWAIT` [mode is not recommended for use and only still exists for backwards compatibility](https://devblogs.microsoft.com/oldnewthing/20110114-00/?p=11753), so it behooves us to pursue an alternate approach.

`WriteFile` on a `PIPE_WAIT` pipe is never guaranteed not to block, so we cannot completely avoid correctness issue 1. without a dedicated writer thread (however, we are confident that writing less data than the free space of a pipe's outbound buffer will not block, even though this is not guaranteed by the API). Because the write is bounded by the size of the outbound buffer in either case, this is unlikely to be a problem.

### Analysis of #853:

In #853, the write-size bounding logic is:
```
if (fpli.WriteQuotaAvailable > 0) {
	len = MIN(len, fpli.WriteQuotaAvailable);
}
len = MIN(len, fpli.OutboundQuota);
```
Keep in mind that `WriteQuotaAvailable = OutboundQuota - space_filled_by_writer - data_requested_by_reader`. So if we have an empty outbound pipe, but a reader in another process requests a full buffer's worth of data, we will counterintuitively see `WriteQuotaAvailable == 0` on the writer (without this idiosyncratic behavior, this problem would be easy to solve).

If a reader is connected to the pipe (`WriteFile` will fail fast if one isn't) but _is not_ attempting to read any data, I suspect that a second write will see `WriteQuotaAvailable == 0` because the write buffer is full, but continue to bound the write size by `OutboundQuota`, and then block the dispatch queue because `WriteFile` will block if the data can't fit in the outbound buffer (correctness issue 1). We can't just skip the write in this case, because `WriteQuotaAvailable` may just be 0 because the buffer is _empty_ but a full buffer's worth of data is being requested by a reader, and then we would never make progress (which would be correctness issue 2).

This sounds bad, however, it just brings the implementation back to how it was before https://github.com/swiftlang/swift-corelibs-libdispatch/pull/781 with the improvement introduced by https://github.com/swiftlang/swift-corelibs-libdispatch/pull/796 to prevent very large writes from overwhelming the queue. As long as a reader is making progress, we will split the write up into `OutboundQuota` sized chunks preventing the dispatch queue from blocking (at least not for long).

**Conclusion: if a reader is connected but not making progress, this solution may block the dispatch queue.**

### Analysis of #854:

In #854, the write-size bounding logic is:
```
if (fpli.WriteQuotaAvailable == 0) {
	len = MIN(len, fpli.OutboundQuota - 1);
} else {
	len = MIN(len, fpli.WriteQuotaAvailable - 1);
}
```
The idea here is that we _never_ write a full buffer's worth of data, which reserves `WriteQuotaAvailable == 0` to indicate that there is a blocked reader waiting for a full buffer's worth of data. When `WriteQuotaAvailable == 1` we assume the pipe is full from the last write, and write 0 bytes, avoiding blocking the dispatch queue. Unfortunately, I suspect there is a possible race condition here which can result in correctness issue 2:
1. The pipe is empty
2. The reader requests exactly `bufsize - 1` bytes of data
3. The writer sees `WriteQuotaAvailable == 1`, assumes the pipe is full, and does not make progress.

I was unable to produce this case in practice, even under synthetic conditions.

This PR also breaks a test, which is why I mentioned a "requirements change" in the PR. There is a test which serially:
1. Writes `bufsize` bytes of data to the pipe
2. Waits for the write to complete
3. Reads `bufsize` bytes of data from the pipe
4. Waits for the read to complete

This no longer works with my PR, because we will only write `bufsize - 1` bytes of data in lieu of a reader making progress. I changed the test to only write/read `bufsize - 1` bytes, and I don't think this is a big deal because there's no reason to serially write and read through a pipe like this on a single thread, and attempting to do this with e.g. `bufsize + 1` bytes was never supported for the same reason. But because this scenario was explicitly tested, I felt it worth calling out as a "requirements change".

**Conclusion: if a reader requests `bufsize - 1` bytes of data from an empty pipe, the write operation may not make progress, but will not block the dispatch queue.**

### Tradeoffs
The current implementation persistently spins the monitoring thread, which does not actually affect correctness, but impacts performance under load and devastates idle performance.

#853 brings the implementation back to how it was 2 years ago, but with some added logic to try and avoid blocking the dispatch queue for too long, but a misbehaving reader may still be able to block the thread.

#854 updates the write bounding logic to guarantee that we avoid blocking the dispatch queue (via writing less than the free space in the buffer; `WriteFile` is never guaranteed to not block), but may introduce a race condition that can prevent a write operation from progressing.

Obviously none of these are perfect, but I think #854 is the least likely to encounter a problem, and the problem that can materialize is often non-fatal (a non-progressing write may be able to be canceled, though not necessarily).

I think that #853 and #854 are both better than the status quo, as the spinning thread affects everyone under every case. This is predicated on my belief that the problems with both solutions are encountered rarely. Of course, if I'm wrong about this, the solutions are not suitable, but I was unable to ever reproduce the problem mentioned with #854, and the problem with #853 lingered in the code for years, and is less likely to be encountered today thanks to https://github.com/swiftlang/swift-corelibs-libdispatch/pull/796.

### Alternatives
As mentioned in the previous comment, the Microsoft-approved way of fixing this will be dedicated writer threads with a signaling mechanism to perform synchronous `WriteFile`s off of the dispatch queue thread in order emulate asynchronous pipe writing semantics. This is likely to impact performance and increase thread-pressure on the system, though it will be correct and much better than the spinning thread of today. This solution also adds complexity, and I'm not sure when I will be able to pursue it.

There's some heuristics we could use to try and avoid the problems mentioned above, but they aren't foolproof. For example: with #854 we could instead write `bufsize - 2` bytes, and then when we see the "pipe full" case (`WriteQuotaAvailable == 2` now) for some period of time (say, 16 iterations of the `dispatch_operation_t` through `_dispatch_operation_perform` with no change), we can try only writing 1 byte. If we were in the ambiguous case where a reader was waiting for bytes, then `WriteQuotaAvailable` will move to `3` and the write will again be able to progress. If after this case, `WriteQuotaAvailable == 1` for a while, we can assume the pipe was actually full and there is no reader making progress, in which case we can take some other action (cancellation, or just waiting forever). But this sort of just feels like kicking the can down the road.

### Conclusion
I didn't intend to write an essay here, my apologies for the long comment 😅 . To directly answer your questions:

> What’s your opinion on which is the better solution to fix this problem? https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 or https://github.com/swiftlang/swift-corelibs-libdispatch/pull/853?

I think #854 is the better approach of the two, although this is predicated on my belief that the incorrectness condition mentioned will be exceedingly rare.

> I might be misunderstanding what you mean by if nothing is actively reading the data. Couldn’t the other end of the pipe be a separate process that just never reads data from the pipe. Since that process might be implemented without libdispatch, I’m not sure how assumptions about the libdispatch reader can affect the writer.

Yes, I think my understanding was off here. The synchronization thread will always signal an I/O completion port when there is data available, but this does not imply that there is an active reader waiting for a signal to start consuming data. If there is a reader connected to the pipe but _not_ reading data, I think that #853 could cause the dispatch queue to become blocked (as explained above).

**Comment by ahoppen** [comment]
Thanks a lot for that detailed compilation of the tradeoff (would have called it summary if not for the length 😅).

I also prefer the solution in #854 because:
1. Having a setup where a reader does not make progress seems like a common mistake that can be made during development. Blocking the writer because of this may make problems like these even harder to diagnose. 
2. The potential error scenario that you described in #854 seems unlikely to occur in practice
3. You weren’t able to produce the correctness issue in #854 in practice
4. I don’t think the change in tests is a blocker since I don’t see a real-world use case where this serial write-read behavior would be required.
5. Both solutions are probably better than the status quo.

Before going ahead, I would like to clarify one thing about your test setup for the potential issue in #854 though:

If you replicate the setup that you describe there, what do you see for `WriteQuotaAvailable`? The test setup sounds to me like it should deterministically produce `WriteQuotaAvailable == 1`, so I’m curious why you can’t reproduce that behavior.

**Comment by z2oh** [comment]
My test setup was attempting to read `bufsize - 1` bytes to trigger the ambiguous case as described, but I never beat the write side to the pipe, so the read would always succeed. I'll try another set of experiments to stress test this and will report back this evening or tomorrow.

**Comment by ahoppen** [comment]
Shouldn’t you be able to hit this pretty consistently by eg. waiting 1 second before doing the write?

**Comment by xwu** [comment]
> There's some heuristics we could use to try and avoid the problems mentioned above, but they aren't foolproof. For example: with #854 we could instead write `bufsize - 2` bytes [...]. But this sort of just feels like kicking the can down the road.

I'm very naive in this area, but in the implementation of our random number generation API, we use a (widely used) rejection sampling technique that _theoretically_ requires an unbounded number of iterations but in practice always terminates very quickly.

Does each heuristic "kick" here move the "can" down the road exponentially further, such that _even if_ you figure out a way to trigger the `bufsize - 1` case consistently, can you effectively make the issue a statistical near-impossibility if (for example) you give yourself three kicks of the can by starting at `bufsize - 4`?

**Comment by z2oh** [comment]
https://github.com/swiftlang/swift-corelibs-libdispatch/pull/854 was merged; I will still follow up here with some additional testing as soon as I can. Apologies for the delay.

> Shouldn’t you be able to hit this pretty consistently by eg. waiting 1 second before doing the write?

I would expect so, but as I recall when I tested this previously in Swift code via the DispatchIO API, the write still went through to the read side after the 1-second delay. That said, I will try this again under a debugger, among some other tests.

> Does each heuristic "kick" here move the "can" down the road exponentially further, such that even if you figure out a way to trigger the `bufsize - 1` case consistently, you effectively make the issue a statistical near-impossibility

This is an interesting idea, but after thinking it through, I believe a non-progressing reader will still quickly reduce us to the base ambiguous case.

To flesh out the algorithm: say the number of bytes to reserve is N, and the number of "stalled iterations" we've undergone so far is M.

We start with M = 0:
1. If we see `WriteQuotaAvailable > (N-M)`, the reader has made progress and we write `WriteQuotaAvailable - N` bytes and set M=0.
2. If we see `WriteQuotaAvailable < (N-M)`, the reader has drained the buffer and requested more data. We write `OutboundQuota - N` bytes and set M=0.
3. If we see `WriteQuotaAvailable == (N-M)`, we are in the ambiguous case: either the reader has made no progress, or the reader has drained the buffer and then raced with the writing thread to request exactly `bufsize - (N-M)` bytes. At this point, if M<(N-1), we do a single byte write and increment M.

If we have an active reader, it's true that the larger N is here, the more "exact bufsize request races" the reader thread would have to win in order to get into a permanently stalled state. But, if we have a non-progressing reader, we will repeatedly hit the ambiguous case, rapidly incrementing M to N and leaving us in the same base ambiguous case as before.

The larger N is, the more confident we can be that M=N was _initially_ caused by a non-progressing reader, but still all it would take is one race win (the reader becoming active and draining the buffer, then racing to request `bufsize - (N-M)` bytes) to get in the same stall situation.

**Comment by DimaKozhinov** [comment]
Swift 6.1.1 is out, but the bug is still there. Do we have to wait for 6.2 release?

**Comment by HeMet** [comment]
> Swift 6.1.1 is out, but the bug is still there. Do we have to wait for 6.2 release?
I downloaded the installer and file version from metadata is `6.1.0`. Maybe links are not updated yet.

**Comment by xwu** [comment]
> Swift 6.1.1 is out, but the bug is still there. Do we have to wait for 6.2 release?

Yes.

**Comment by bnbarham** [comment]
Re-opening since it's been reverted from main (and being reverted from 6.2) as well.

**Comment by roman-bcny** [comment]
Further investigation showed PIPE_WAIT hang report resulting in the revert (https://github.com/swiftlang/swift-corelibs-libdispatch/issues/882) was caused by Read+Peek, not Write calls, and there doesn't seem to be an easy way to get around this on libdispatch side. Details and trying to find a way forward here: https://forums.swift.org/t/sourcekit-lsp-cpu-spinning-and-libdispatch-complications/82472

**Comment by DimaKozhinov** [comment]
Closing this issue, as it has been fixed in Swift 6.4.

**Comment by ahoppen** [comment]
The underlying issue in corelibs-libdispatch still exists. SourceKit-LSP worked around it by no longer using corelibs-libdispatch to read from stdin.
